### PR TITLE
ARROW-1993: [Python] Add function for determining implied Arrow schema from pandas.DataFrame

### PR DESCRIPTION
**Pull request by keechongtan**

### CODE REVIEW

**Comment by xhochy** [comment]
cc @fjetter and @crepererum: This might be interesting for you. This is a first iteration of extracting Arrow schemas from Pandas DataFrames without doing the full conversion. It still does the conversion on `object` columns but I would add a follow-up PR that exposes some of the type guessing functions from C++ here so that we can call them instead of the full conversion.

**Comment by wesm** [comment]
@keechongtan could you rebase this? I can try to take a closer look soon

**Comment by keechongtan** [comment]
@wesm rebased.

**Comment by pitrou** [comment]
Hi @keechongtan, thanks for posting the PR! I haven't taken a detailed look at your code yet, but there seem to be some problems (perhaps due to rebasing?):
- the `flake8` step fails on Travis-CI
- the `_ndarray_to_array` and `from_pandas` functions seem duplicated
- Cython compilation fails (both on AppVeyor and Travis-CI)

**Comment by kszucs** [comment]
@pitrou @wesm I made this PR up to date. 

I intend to refactor the pandas_compat module afterwards, because it's pretty hard to follow what happens there.

**Comment by wesm** [comment]
I will have a look

File: python/pyarrow/pandas_compat.py [hunk 1]
```diff
@@ -365,6 +364,44 @@ def dataframe_to_arrays(df, schema, preserve_index, nthreads=1, columns=None,
...
+    for c in columns_to_convert:
+        values = c.values
+        if isinstance(c.values, pd.Categorical):
+            type_ = pa.array(c, from_pandas=True).type
+        else:
+            values, type_ = get_datetimetz_type(values, c.dtype, None)
+            type_ = pa.lib._ndarray_to_arrow_type(values, type_)
+            if type_ is None:
+                type_ = pa.array(c, from_pandas=True).type
+        types.append(type_)
```
Comment (wesm) [inline python/pyarrow/pandas_compat.py]: Well, this is certainly not cheap to compute in all cases. I think it's fine to focus on getting the API and behavior correct, then focusing later on performance

File: python/pyarrow/types.pxi [hunk 2]
```diff
@@ -591,6 +591,45 @@ cdef class Schema:
...
+    @classmethod
+    def from_pandas(cls, df, bint preserve_index=True):
+        """
+        Returns implied schema from dataframe
+
+        Parameters
+        ----------
+        df : pandas.DataFrame
+        preserve_index : bool, optional
+            Whether to store the index as an additional column in the resulting
```
Comment (wesm) [inline python/pyarrow/types.pxi]: Maybe add "(or columns, for MultiIndex)"

File: python/pyarrow/types.pxi [hunk 3]
```diff
@@ -591,6 +591,45 @@ cdef class Schema:
...
+        >>> import pandas as pd
+        >>> import pyarrow as pa
+        >>> df = pd.DataFrame({
+            ...     'int': [1, 2],
+            ...     'str': ['a', 'b']
+            ... })
+        >>> pa.Schema.from_pandas(df)
+        int: int64
+        str: string
+        __index_level_0__: int64
```
Comment (wesm) [inline python/pyarrow/types.pxi]: Do you want to add an example of this to the Sphinx documentation?

**Review by wesm — COMMENTED** [review]
Looks good. Left some comments, mostly around documentation

File: python/pyarrow/pandas_compat.py (hunk 1 above)
Comment (kszucs) [reply python/pyarrow/pandas_compat.py]: Yeah, my intention is to finalize this PR and improve it later.

File: python/pyarrow/types.pxi (hunk 3 above)
Comment (kszucs) [reply python/pyarrow/types.pxi]: pandas.rst lacks of examples, so just mentioned it :)
